### PR TITLE
Environments: unify the spec objects on first concretization

### LINKED ISSUE #29947: Traversing specs in the root direction doesn't work if an environment is not loaded from spack.lock
### Steps to reproduce

Using this environment:
```yaml
spack:
  specs:
  - hdf5~mpi
  - hdf5+mpi
```
and this simple `script.py`:
```python
import spack.ci
import spack.environment

e = spack.environment.Environment('/tmp/ci')
e.concretize()
# e.write()

for item in spack.ci.get_spec_filter_list(e, ['zlib']):
    print(item)
```
shows a lot less specs than expected if a `spack.lock` is not present. With a `spack.lock` the result is correct.

### Error message

This is the result we get:
```console
$ spack python script.py 
==> Starting concretization pool with 2 processes
==> Environment concretized in 8.00 seconds.
zlib@1.2.12%gcc@9.4.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-icelake
hdf5@1.12.1%gcc@9.4.0~cxx~fortran~hl~ipo~java~mpi+shared~szip~threadsafe+tools api=default build_type=RelWithDebInfo patches=ee351eb arch=linux-ubuntu20.04-icelake ^cmake@3.23.0%gcc@9.4.0~doc+ncurses+ownlibs~qt build_type=Release arch=linux-ubuntu20.04-icelake ^ncurses@6.2%gcc@9.4.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-icelake ^openssl@1.1.1n%gcc@9.4.0~docs~shared certs=system arch=linux-ubuntu20.04-icelake ^pkgconf@1.8.0%gcc@9.4.0 arch=linux-ubuntu20.04-icelake ^zlib@1.2.12%gcc@9.4.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-icelake
```
The error is that, when we concretize the specs separately in memory we have separate Python objects for specs which share the same hash. That means each of the objects with a given hash has only a part of the references towards the root specs.

With a `spack.lock` we have instead a single Python object per spec hash, and so traversing towards roots works correctly.

### Information on your system

* **Spack:** 0.18.0.dev0 (b2eda32e55601dd80bd0602527e5736e24ef8544)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

*This error occurs when pruning specs in Gitlab Pipelines*

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by alalazo**
fixes #29947 

Modifications:
- [x] Do a roundtrip to/from lockfile dict in memory to unify the spec objects
- [x] Add unit test for regression
- [x] Make pointless modifications to `mpich` to trigger jobs in Gitlab Pipelines

### CODE REVIEW

**Review by raffenet — DISMISSED** [review]

**Review by scottwittenburg — COMMENTED** [review]
Looks good to me @alalazo, thanks for tackling this so quickly! 

You have one approval already, but I'm happy to do so again once we see the results in the gitlab pipeline, and once you've had a chance to remove the pointless `mpich` commit.

**Comment by raffenet** [comment]
So the MPICH commit is just a dummy commit to trigger a pipeline? I guess we (MPICH maintainers) can ignore it then.

**Comment by raffenet** [comment]
> So the MPICH commit is just a dummy commit to trigger a pipeline? I guess we (MPICH maintainers) can ignore it then.

Oh, I see it now in the PR description.

**Comment by scottwittenburg** [comment]
> So the MPICH commit is just a dummy commit to trigger a pipeline? I guess we (MPICH maintainers) can ignore it then.

Yes, sorry about that.  Since you mention it, I'll just provide some explanation.  You were automatically added as a reviewer by spackbot when it noticed the `package.py` was touched and it saw you were listed as a maintainer.  

Why did we have to touch it?  Since we merged #29697 to reduce unnecessary builds on PRs, we've started adding temporary commits touching packages to actually run gitlab builds for "core-only changes".   In this case, we specifically needed to test what happened when `mpich` was touched, because we discovered a bug where a PR (one of yours actually, #28573) touched it, but then gitlab erroneously pruned jobs for its dependents, allowing a breakage into `develop`.

Maybe we can come up with a better way to force gitlab testing when packages aren't touched, or maybe we can adopt the convention of adding `WIP` in the commit message and try to have spackbot ignore those when adding reviewers.

**Comment by scottwittenburg** [comment]
@zackgalbreath this fixes the excessive pruning issue we discussed yesterday.

File: var/spack/repos/builtin/packages/mpich/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/mpich/package.py:28]: unrelated changes?

File: lib/spack/spack/environment/environment.py
Comment (tldahlgren) [inline lib/spack/spack/environment/environment.py:1277]: LGTM.  New and modified tests pass for me locally.

Is there a reason for using the `build_hash` versus others?  If so, perhaps it would be good to explain the hash choice in a comment.

Comment (alalazo) [reply lib/spack/spack/environment/environment.py:1277]: Currently environments are indexed by build hashes. When looking into this bug I noticed there is a disconnect between environments that are concretized in memory for the first time and environments that are read from a `spack.lock`. The issue is that specs read from a `spack.lock` don't have a full hash, since they are indexed by a build hash which is strictly coarser. They are also marked "final" as they are read from a file, so we can't compute additional hashes. 

This bugfix PR makes "first concretization" equivalent to re-reading the specs from a corresponding `spack.lock`, and doing so unveiled a few tests where we were making wrong assumptions and relying on the fact that a `spack.lock` file was not there already.

**Comment by alalazo** [comment]
@spackbot run pipeline

**Review by tgamblin — APPROVED** [review]

**Pull request merged**
